python/pyspark/sql/streaming/listener.py

```diff
@@ -155,11 +156,31 @@ class QueryStartedEvent:
     This API is evolving.
     """
 
-    def __init__(self, jevent: JavaObject) -> None:
-        self._id: uuid.UUID = uuid.UUID(jevent.id().toString())
-        self._runId: uuid.UUID = uuid.UUID(jevent.runId().toString())
-        self._name: Optional[str] = jevent.name()
-        self._timestamp: str = jevent.timestamp()
+    def __init__(
+        self, id: uuid.UUID, runId: uuid.UUID, name: Optional[str], timestamp: str
+    ) -> None:
+        self._id: uuid.UUID = id
+        self._runId: uuid.UUID = runId
+        self._name: Optional[str] = name
+        self._timestamp: str = timestamp
+
+    @classmethod
+    def fromJObject(cls, jevent: JavaObject) -> "QueryStartedEvent":
+        return cls(
+            id=uuid.UUID(jevent.id().toString()),
+            runId=uuid.UUID(jevent.runId().toString()),
+            name=jevent.name(),
+            timestamp=jevent.timestamp(),
+        )
+
+    @classmethod
+    def fromJson(cls, j: Dict[str, Any]) -> "QueryStartedEvent":
```

Comment:
What is the context where these are used? Thanks for the detailed description and you mentioned it for 'step 3'. Could you point to any WIP code that uses this?

Reply:
Sure! Here is the pointer: https://github.com/apache/spark/pull/41096/files#r1230025798
Note that it's draft code so it's really messy. Basically since we register a scala listener but run python code inside it. We need a way to send the events to python process. And JSON is a safe way for that job

Reply:
I see. Scala does toJson and Python code does fromJson. 
Should this be under connect directory since it is Connect specific code?

Reply:
I think we need to still let them use `import pyspark.sql.streaming.Query<xxx>Event` in Connect. If we put the code in connect folder they might need to do `import pyspark.sql.connect.streaming.Query<xxx>Event`.

A way I can think of doing this is to refactor the code to make the event classes abstract here:
```
class QueryStartedEvent(ABC):
    @property
    @abstractmethod
    def id(self) -> uuid.UUID:
        ...
```

Do you think we should do that?

Reply:
> I think we need to still let users to use import pyspark.sql.streaming.Query<xxx>Event

@HyukjinKwon is this true? I think Users have only the Spark Connect python code. 

> import pyspark.sql.connect.streaming.Query<xxx>Event.

User code in a StreamingListener in connect is a Spark Connect code. I think our imports statements import the right version (connect or legacy depending on the environment).

Reply:
Yeah, actually I didn't want users to directly improve the event class (and that's why I mentioned that the constructor is private). Do we need for end users to be able to import in Spark Connect?

Reply:
They would write StreamingListener code just like their rest of Spark-Connect code. I.e. they have access to only those packages that are available in Spark-Connect. (they don't import 'connect' version directly).

What that would mean for this class is a bit uncertain to me. If @HyukjinKwon is ok with this, I am ok.